### PR TITLE
ISPN-8729 Use async operations in Hot Rod server

### PR DESCRIPTION
**Pull request by rvansa**
https://issues.jboss.org/browse/ISPN-8729

Replaces #5706 Let's see if CI behaves sane here...

### CODE REVIEW

**Comment by rvansa** [comment]
@tristantarrant Only unrelated test failures. Seems that once the PR gets CI upset there's no way back...

Review now or forever hold your peace.

**Comment by gustavonalle** [comment]
Triggering another CI

**Comment by gustavonalle** [comment]
Looking

File: commons/src/main/java/org/infinispan/commons/api/AsyncCache.java [hunk 1]
```diff
@@ -265,20 +269,37 @@
...
+   /**
+    * Asynchronous version of {@link BasicCache#containsKey(Object)}
+    * @param key key to retrieve
+    * @return future containing true if the mapping exists.
+    */
+   default CompletableFuture<Boolean> containsKeyAsync(K key) {
+      return getAsync(key).thenApply(Objects::nonNull);
+   }
+
+   default CompletableFuture<Map<K, V>> getAllAsync(Set<?> keys) {
```
Comment (galderz) [inline commons/src/main/java/org/infinispan/commons/api/AsyncCache.java]: I was just wondering about this yesterday. We had a `putAll` but there was no `getAll` 👍

**Comment by galderz** [comment]
Looking at this this morning

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/BaseRequestProcessor.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
+package org.infinispan.server.hotrod;
+
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.Executor;
+
+import io.netty.channel.Channel;
+
+/**
+ * This class is not thread safe. The expected use is to set {@link #cdc}, call some method in inheriting classes
```
Comment (galderz) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/BaseRequestProcessor.java]: Link should be fixed?

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+   private static AdvancedCache<byte[], byte[]> unsetSubject(AdvancedCache<byte[], byte[]> cache) {
+      if (cache instanceof SecureCacheImpl) {
+         return new SecureCacheImpl<>(SecurityActions.getUnwrappedCache(cache));
+      } else {
+         return cache;
+      }
+   }
+
+   private boolean isBlockingRead(CacheDecodeContext cdc, CacheInfo info) {
+      return info.persistence && !cdc.decoder.isSkipCacheLoad(cdc.header);
```
Comment (galderz) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java]: Why are the read/writes blocking when there's persistence? Is it because we don't have async drivers for all types of persistence that can be attached? Or is it because we don't yet have async API for persistence layer within Infinispan?

File: server/hotrod/src/test/java/org/infinispan/server/hotrod/HotRodMergeTest.java [hunk 4]
```diff
@@ -73,6 +73,7 @@ protected void createCacheManagers() {
       waitForClusterToForm();
    }
 
+   @Test(enabled = false)
```
Comment (galderz) [inline server/hotrod/src/test/java/org/infinispan/server/hotrod/HotRodMergeTest.java]: Why are these disabled? There's no comment/description. Make sure you create a follow up JIRA to re-enable them or delete them if not relevant any more.

File: server/integration/jgroups/src/main/resources/subsystem-templates/infinispan-jgroups.xml [hunk 5]
```diff
@@ -19,8 +19,8 @@
                 <protocol type="UNICAST3" />
                 <protocol type="pbcast.STABLE" />
                 <protocol type="pbcast.GMS" />
-                <protocol type="UFC" />
-                <protocol type="MFC" />
+                <protocol type="UFC_NB" />
```
Comment (galderz) [inline server/integration/jgroups/src/main/resources/subsystem-templates/infinispan-jgroups.xml]: Are these changes related to this JIRA?

**Comment by galderz** [comment]
Btw @rvansa, I guess we can close [ISPN-5083](https://issues.jboss.org/browse/ISPN-5083) when this is in :)

**Comment by galderz** [comment]
Also, is [ISPN-6471](https://issues.jboss.org/browse/ISPN-6471) still relevant?

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java (hunk 3 above)
Comment (tristantarrant) [reply server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java]: Tha latter

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/BaseRequestProcessor.java (hunk 2 above)
Comment (rvansa) [reply server/hotrod/src/main/java/org/infinispan/server/hotrod/BaseRequestProcessor.java]: Oh, the whole doc is wrong... At first I didn't realize that the CacheDecodeContext can be accessed by a different thread later on, so I've reused it but I had to revert this. Good catch.

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java (hunk 3 above)
Comment (rvansa) [reply server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java]: Yes, the persistence API does not support async approach, so even if there was an async driver we could not use it.
There are several layers where we could address it. For simplicity I did the checks here despite it's the least effective. I could also let persistence interceptors do invoke the synchronous API from an executor (if needed*), which would give the non-blocking behaviour to embedded mode as well. But the ideal situation would be an async persistence API.

* We could introduce Flag.HINT_SYNC that would be on for the regular `put()` calls and would skip the delegation to another thread pool when the thread is about to wait anyway.

File: server/integration/jgroups/src/main/resources/subsystem-templates/infinispan-jgroups.xml (hunk 5 above)
Comment (rvansa) [reply server/integration/jgroups/src/main/resources/subsystem-templates/infinispan-jgroups.xml]: Yes, removing blocking in JGroups is another step in the async crusade.

**Comment by rvansa** [comment]
@galderz Yes, I didn't find ISPN-5083. I've closed ISPN-6471 as that was incorporated in Will's refactoring.

**Comment by rvansa** [comment]
@galderz Updated.

File: commons/src/main/java/org/infinispan/commons/api/AsyncCache.java [hunk 6]
```diff
@@ -265,20 +269,37 @@
...
+      return CompletableFuture.completedFuture(
+            keys.stream().map(k -> new AbstractMap.SimpleEntry<>((K) k, getAsync((K) k)))
+               .collect(Collectors.toList()).stream()
```
[the remark this answers is not in the document]
Comment (rvansa) [reply commons/src/main/java/org/infinispan/commons/api/AsyncCache.java]: @gustavonalle To make the invocation really parallel. I need to force all the `getAsync` to be called before calling the first `join`. On the other hand I don't need to return the completedFuture; I'll rework this.
It makes sense to add some tests; do you want me to test the default implementation, too?

File: core/src/main/java/org/infinispan/AdvancedCache.java [hunk 7]
```diff
@@ -386,6 +423,22 @@
...
+    * An overloaded form of {@link #replaceAsync(K, V)}, which takes in an instance of {@link Metadata} which can be used to
+    * provide metadata information for the entry being stored, such as lifespan, version of value...etc. The {@link
+    * Metadata} is only stored if the call is successful.
+    *
+    * @param key      key with which the specified value is associated
+    * @param value    value to be associated with the specified key
+    * @param metadata information to store alongside the new value
+    * @return the future that contains previous value associated with the specified key, or <tt>null</tt>
+    *    if there was no mapping for the key.
+    * @since 9.3
```
[the remark this answers is not in the document]
Comment (rvansa) [reply core/src/main/java/org/infinispan/AdvancedCache.java]: Tristan keeps postponing, I thought there's not chance to get it ready in time :)

File: server/integration/jgroups/src/main/resources/subsystem-templates/infinispan-jgroups.xml [hunk 8]
```diff
@@ -19,8 +19,8 @@
...
+                <protocol type="MFC_NB" />
```
[the remark this answers is not in the document]
Comment (rvansa) [reply server/integration/jgroups/src/main/resources/subsystem-templates/infinispan-jgroups.xml]: Good catch.

**Comment by rvansa** [comment]
@gustavonalle Good that you've requested the test; I've missed overriding in DecoratedCache && there was even a problem in BaseDistributionInterceptor which I had to fix.

**Comment by tristantarrant** [comment]
There is an NPE in https://ci.infinispan.org/job/Infinispan/job/PR-5749/4/testReport/org.infinispan.server.test.task/DistributedServerTaskIT/shouldPutNewValueInRemoteCache/

It seems slightly different from previos failures of this test and related to tx synchronization. Could the async stuff affect this ?

**Comment by rvansa** [comment]
@tristantarrant It is triggered by this PR (using putAsync on tx cache) but the bug is elsewhere.
[ISPN-8853 Registering interposed synchronization fails in async operation with implicit transaction](https://issues.jboss.org/browse/ISPN-8853)

**Comment by tristantarrant** [comment]
Merged. Thanks
